splicemachine/spliceengine
# DB-11841 Log4j version 2 + Row Trigger Performance Improvements

### The following PR in ee should be merged along with this one:  https://github.com/splicemachine/spliceengine-ee/pull/652


## Short Description
This updates Log4j in Splice from version 1 to version 2 and reduces the CPU path length for row triggers.

## Long Description
Profiling of trigger execution paths was carried out to identify bottlenecks in need of correction.  Statement logging turned out to be a critical bottleneck, so Log4j is upgraded from version 1 to version 2 in order to pick up asychronous logger support.  The main difference in Log4j version 2 that everyone needs to know about is how to get a logger.  
In version 1 we use Logger.getLogger, but in version 2 we use LogManager.getLogger.
version 1:
> private static final Logger LOG = Logger.getLogger(GenericPreparedStatement.class);

version 2:
> private static final Logger LOG = LogManager.getLogger(GenericPreparedStatement.class);

Also, all version 2 log4j imports begin with **import org.apache.logging**, such as:
> import org.apache.logging.log4j.LogManager;
> import org.apache.logging.log4j.Logger;

Going forward, please **do not** include version 1 imports, such as:
> import org.apache.log4j.Logger;

The log4j dependency is only defined in one pom file now, utilities/pom.xml.  Please do not add direct dependencies on log4j going forward, but add the following dependency instead:

>         <dependency>
>             <groupId>com.splicemachine</groupId>
>             <artifactId>utilities</artifactId>
>             <version>${project.version}</version>
>         </dependency>

Changes done:

- Include the log4j and scala dependencies in one location only: utilities/pom.xml
- Define ${log4j.version} and ${scala.version} in the main pom file.
- Exclude different versions of log4j and scala brought in by dependencies.
- Rename info-log4j.properties to info-log4j2.properties.
- Update log4j properties files to use the new syntax (see "Configuration with Properties" at https://logging.apache.org/log4j/log4j-2.4/manual/configuration.html)
- Update log file path names to use sys property lookups (see https://logging.apache.org/log4j/2.x/manual/lookups.html)
- Use asychronous loggers on standalone and physical clusters by configuring "Log4jContextSelector=org.apache.logging.log4j.core.async.AsyncLoggerContextSelector" in appropriate locations, see https://logging.apache.org/log4j/2.x/manual/async.html.  Asychronous logging on cloud is working with the latest version of the code.
- Fix join strategy hinting so it does not apply to single-table scans.
- Make subqueries non-materializable (for correctness) if they reference the trigger NEW or OLD rows.
- Remove the rowConcurrentExecutorsMap dependence on TriggerDescriptor

### Performance Improvements
- Asychronous logging removes a large bottleneck concurrent triggers encouter due to simultaneous statement logging.
- Add a protocolKeyCache to BaseMonitor
- Add caching of ScanOperations used inside a trigger as long as there is no DML statement in the trigger and no non-deterministic functions.
- Avoid pushing nondeterministic predicates to the base table in a trigger, to allow the base table rows to be cached. 
- Always compile row triggers to run in OLTP mode.
- Only build the trigger WHEN clause subquery operation tree on the first execution.
- Change logStartExecuting and logEndExecuting to use StringBuilder for constructing the statement string to log.

## How to test
A benchmark for concurrent row triggers is added: https://github.com/splicemachine/spliceengine/blob/d8174af313fd7068117839cecb8feea64efcad87/splice_machine/src/test/java/com/splicemachine/benchmark/ConcurrentTriggerBenchmark.java

For feature testing, some test combinations to try are:

1. WHEN clause with different types of subqueries, referencing correlated columns, the NEW and OLD trigger row, and non-deterministic functions
2. Triggers with the SET clause and/or SIGNAL statement
3. Triggers which hint spark (all row triggers should now execute in control if the main DML executes in control)
4. Multiple actions in a row trigger
5. Row triggers which cause other row triggers or statement triggers to fire
